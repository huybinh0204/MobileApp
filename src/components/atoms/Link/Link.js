import PropTypes from 'prop-types';
import React from 'react';
import BaseText from '_components/atoms/BaseText/BaseText';
import { LINK_TYPES } from '_constants';
import { openExternalLink } from '_utilities/ExternalLinks';
import LINK_URL_PREFIXES from './LinkUrlPrefixes';

const Link = ({ type, href, children, wrapperComponent, ...wrapperProps }) => {
  const Wrapper = wrapperComponent ?? BaseText;

  const onLinkPress = async () => {
    const link = `${LINK_URL_PREFIXES[type]}${href}`;
    await openExternalLink(link);
  };

  return (
    <Wrapper onPress={onLinkPress} {...wrapperProps}>
      {children}
    </Wrapper>
  );
};

Link.defaultProps = {
  type: LINK_TYPES.WEB,
  wrapperComponent: null,
  wrapperProps: null,
};

Link.propTypes = {
  type: PropTypes.oneOf(Object.values(LINK_TYPES)),
  href: PropTypes.string.isRequired,
  children: PropTypes.string.isRequired,
  wrapperProps: PropTypes.object,
};

export default Link;
