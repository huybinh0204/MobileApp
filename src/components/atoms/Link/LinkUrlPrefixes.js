export default {
  web: '',
  sms: 'sms:',
  phone: 'tel:',
  mail: 'mailto:',
};
