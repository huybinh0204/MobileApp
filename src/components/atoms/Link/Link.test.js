import { fireEvent, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import { Linking } from 'react-native';
import { LINK_TYPES } from '_constants';
import Theme from '_utilities/Theme';
import Link from './Link';
import LINK_URL_PREFIXES from './LinkUrlPrefixes';

const TEST_LINK_MAP = {
  WEB: {
    props: {
      type: LINK_TYPES.WEB,
      href: 'lendup.com',
      children: 'my-web-address',
    },
    urlPrefix: LINK_URL_PREFIXES.web,
  },
  SMS: {
    props: {
      type: LINK_TYPES.SMS,
      href: '1-855-253-6387',
      children: 'my-mobile-number',
    },
    urlPrefix: LINK_URL_PREFIXES.sms,
  },
  PHONE: {
    props: {
      type: LINK_TYPES.PHONE,
      href: '1-855-253-6387',
      children: 'my-phone-number',
    },
    urlPrefix: LINK_URL_PREFIXES.phone,
  },
  MAIL: {
    props: {
      type: LINK_TYPES.MAIL,
      href: 'someaddress@lendup.com',
      children: 'my-email-address',
    },
    urlPrefix: LINK_URL_PREFIXES.mail,
  },
};

describe('Link', () => {
  it('should test all link types', () => {
    expect(Object.keys(TEST_LINK_MAP).sort()).toEqual(Object.keys(LINK_TYPES).sort());
  });

  it('should match the snapshot', () => {
    const { toJSON } = render(
      <Theme>
        <Link {...TEST_LINK_MAP.PHONE.props} />
      </Theme>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  let testLinkType = (link) => {
    it(`should call canOpenURL when pressed for ${link.props.type}`, async () => {
      jest.spyOn(Linking, 'canOpenURL');
      jest.spyOn(Linking, 'openURL');

      const { getByText } = render(
        <Theme>
          <Link {...link.props} />
        </Theme>
      );

      const LinkButton = getByText(link.props.children);

      const expectedLinkUrl = `${link.urlPrefix}${link.props.href}`;

      fireEvent.press(LinkButton);

      await waitFor(() => {
        expect(Linking.canOpenURL).toHaveBeenCalledWith(expectedLinkUrl);
        expect(Linking.openURL).toHaveBeenCalledWith(expectedLinkUrl);
      });
    });
  };

  for (const name in TEST_LINK_MAP) {
    testLinkType(TEST_LINK_MAP[name]);
  }
});
