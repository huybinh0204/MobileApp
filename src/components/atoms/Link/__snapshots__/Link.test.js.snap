// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should match the snapshot 1`] = `
<Text
  allowFontScaling={false}
  onPress={[Function]}
  wrapperProps={null}
>
  my-phone-number
</Text>
`;
